src/undo.js

```diff
@@ -14,6 +14,108 @@ import ButtonView from '../ui/button/buttonview.js';
+ * ready to be applied. In our scenario, delta `C3` is the last delta so `C3r` bases on up-to-date document state so
+ * it can be applied to the document.
+ *
+ *		  history                           undo stack
+ *		===========             ==================================
+ *		[delta A1 ]             [batch A with selection before A1]
+ *		[delta B1 ]             [batch B with selection before B1]
+ *		[delta B2 ]             [   processing undoing batch C   ]
+ *		[delta C1 ]
+ *		[delta C2 ]
+ *		[delta B3 ]
+ *		[delta C3 ]
+ *		[delta C3r]
+ *
+ * Next is delta `C2`, reversed to `C2r`. `C2r` bases on `C2`, so it bases on wrong document state. It needs to be
+ * transformed by deltas from history that happened after it, so it "knows" about them. Let `C2' = C2r * B3 * C3 * C3r`,
+ * where `*` means "transformed by". As can be seen, `C2r` is transformed by a delta which is undone afterwards anyway.
+ * This brings two problems: lower effectiveness (obvious) and incorrect results. Bad results come from the fact that
+ * operational transformation algorithms assume there is no connection between two transformed operations when resolving
+ * conflicts, which is true for, i.e. collaborative editing, but is not true for undo algorithm.
+ *
+ * To prevent both problems, `CompressedHistory` introduces an API to {@link engine.model.CompressedDelta#removeDelta remove}
+ * deltas from history. It is used to remove undone and undoing deltas after they are applied. It feels right - delta is
+ * undone/reversed = "removed", there should be no sign of it in history (fig. 1). `---` symbolizes removed delta.
+ *
+ *		history (fig. 1)            history (fig. 2)            history (fig. 3)
+ *		================            ================            ================
+ *		   [delta A1]                  [delta A1]                  [delta A1]
+ *		   [delta B1]                  [delta B1]                  [delta B1]
+ *		   [delta B2]                  [delta B2]                  [delta B2]
+ *		   [delta C1]                  [delta C1]                  [---C1---]
+ *		   [delta C2]                  [---C2---]                  [---C2---]
+ *		   [delta B3]                  [delta B3]                  [delta B3]
+ *		   [---C3---]                  [---C3---]                  [---C3---]
+ *		   [---C3r--]                  [---C3r--]                  [---C3r--]
+ *		                               [---C2'--]                  [---C2'--]
+ *		                                                           [---C1'--]
+ *
+ * Now we can transform `C2r` only by `B3` and remove both it and `C2` (fig. 2). Same with `C1` (fig. 3). `'` symbolizes
+ * reversed delta that was later transformed.
+ *
+ * But what about that selection? For batch `C`, undo feature remembers selection just before `C1` was applied. It can be
+ * visualized between delta `B2` and `B3` (see fig. 3). Obviously a lot happened to the document since the selection
```

Comment:
Maybe "a lot" is overstating here, yes. But what I've meant is that we actually applied five deltas. But it's true that for that selection only `B3` happens. I'll change it.